### PR TITLE
Don't allow incompatible keywords in `DataFrame.dropna`

### LINKED ISSUE #9365: Fix  `DataFrame.dropna` match pandas upstream bug fix
Currently, these  two UPSTREAM test are broken:

- [test_map_partitions_with_delayed_collection](https://github.com/dask/dask/runs/7331385905?check_suite_focus=true#step:5:11853)
- [test_dropna](https://github.com/dask/dask/runs/7331385905?check_suite_focus=true#step:5:11560)

Due to a [BUG-fix upstream](https://github.com/ncclementi/pandas/commit/94280f021bda7fb23e26db6e485e5367f2de8963), that prevents providing `how` and `thresh` simultaneously to `pd.DataFrame.dropna()` . The change involves modifying the default values `how='any', thresh=None,` to `how=_NoDefault.no_default, thresh=_NoDefault.no_default,` (see https://pandas.pydata.org/docs/dev/reference/api/pandas.DataFrame.dropna.html ) where` _NoDefault.no_default` comes from https://github.com/pandas-dev/pandas/blob/e7afa4b641b146874d17c36caa8a050bfde31283/pandas/_libs/lib.pyi#L27-L31

### PR DESCRIPTION
**Pull request by ncclementi**
- [x] Closes #9365
- [x] Tests added / passed
- [x] Passes `pre-commit run --all-files`

### CODE REVIEW

File: dask/dataframe/core.py [hunk 1]
```diff
@@ -5054,8 +5054,17 @@ def eval(self, expr, inplace=None, **kwargs):
         meta = self._meta.eval(expr, inplace=inplace, **kwargs)
         return self.map_partitions(M.eval, expr, meta=meta, inplace=inplace, **kwargs)
 
+    if PANDAS_GT_150:
+        from pandas._libs.lib import no_default
```
Comment (ian-r-rose) [inline dask/dataframe/core.py:5058]: Nit: rather than importing from a private module, we can import from `pandas.api.extensions`. It's not a huge deal, and pandas doesn't provide very strong guarantees around this, it seems, but see some discussion in https://github.com/pandas-dev/pandas/issues/40397

File: dask/dataframe/core.py [hunk 2]
```diff
@@ -5054,8 +5054,17 @@ def eval(self, expr, inplace=None, **kwargs):
         meta = self._meta.eval(expr, inplace=inplace, **kwargs)
         return self.map_partitions(M.eval, expr, meta=meta, inplace=inplace, **kwargs)
 
+    if PANDAS_GT_150:
```
Comment (ian-r-rose) [inline dask/dataframe/core.py:5057]: This clause is kind of odd, in that it attaches `how_default` and `thresh_default` to `_Frame` as class variables, which I don't think is what you intended. Probably cleaner to keep this logic at the module scope.

**Review by ian-r-rose — COMMENTED** [review]
Thanks @ncclementi.

One of these days, I think we should reconsider some of the policies around all of these backwards incompatibility checks, and probably phase some of them out. But this approach seems fine for now.

**Comment by jsignell** [comment]
> One of these days, I think we should reconsider some of the policies around all of these backwards incompatibility checks, and probably phase some of them out. 

This has been niggling at me for a while. It would be good to create a policy for how Dask should handle this kind of API change.

File: dask/dataframe/core.py (hunk 1 above)
Comment (jrbourbeau) [reply dask/dataframe/core.py:5058]: +1

File: dask/dataframe/core.py [hunk 3]
```diff
@@ -5054,8 +5054,17 @@ def eval(self, expr, inplace=None, **kwargs):
...
+
+        how_default = no_default
+        thresh_default = no_default
+    else:
+        how_default = "any"
+        thresh_default = None
+
     @derived_from(pd.DataFrame)
-    def dropna(self, how="any", subset=None, thresh=None):
+    def dropna(self, how=how_default, subset=None, thresh=thresh_default):
```
Comment (jrbourbeau) [inline dask/dataframe/core.py:5067]: Currently we're trying to match `pandas` behavior here exactly, which is normally a guiding philosophy. However, in this particular case, since the change in `pandas` was to fix a bug where incompatible keywords could be specified and unexpected results returned https://github.com/pandas-dev/pandas/issues/46575, I think it would make sense for us to always raise an error when `how` and `thresh` are both specified (i.e. always match the new `pandas` behavior). While this does lead to slightly different behavior for older versions of  `pandas`, Dask users would still benefit from not having an unexpected result returned

File: dask/dataframe/tests/test_dataframe.py [hunk 4]
```diff
@@ -762,8 +762,8 @@ def test_dropna():
     )
 
     # threshold
-    assert_eq(df.dropna(thresh=None), df.loc[[20, 40]])
-    assert_eq(ddf.dropna(thresh=None), df.dropna(thresh=None))
+    assert_eq(df.dropna(), df.loc[[20, 40]])
+    assert_eq(ddf.dropna(), df.dropna())
```
Comment (jrbourbeau) [inline dask/dataframe/tests/test_dataframe.py:766]: Hrm, why did we drop `thresh=` here? It looks like it wasn't being used in conjunction with `how=` in this case (though I could be missing something)

Comment (jrbourbeau) [reply dask/dataframe/tests/test_dataframe.py:766]: Could you also add a test that checks an error is raised when both `how` and `thresh` are specified?

File: dask/dataframe/core.py (hunk 2 above)
Comment (ncclementi) [reply dask/dataframe/core.py:5057]: whoops, yes you are right, I'll leave outside the class.

File: dask/dataframe/tests/test_dataframe.py (hunk 4 above)
Comment (ncclementi) [reply dask/dataframe/tests/test_dataframe.py:766]: I think I was trying something and I forgot, and now I see that if they are there, we have an AssertionError for 
` assert_eq(df.dropna(thresh=None), df.loc[[20, 40]])` I have to take a deeper look. I'll push the code so it's visible.

File: dask/dataframe/core.py (hunk 3 above)
Comment (ncclementi) [reply dask/dataframe/core.py:5067]: @jrbourbeau Should we then set both to `no_default` and handle manually the case pandas < 1.5. following what pandas does, something like

```python
    @derived_from(pd.DataFrame)
    def dropna(self, how=how_default, subset=None, thresh=thresh_default):
        if not PANDAS_GT_150 and how is not no_default and thresh is not no_default:
            raise ValueError(
                "You cannot set both the how and thresh arguments at the same time."
            )

         if how is no_default:
            how = "any"  
          
        return self.map_partitions(
            M.dropna, how=how, subset=subset, thresh=thresh, enforce_metadata=False
        )
```

File: dask/dataframe/core.py [hunk 5]
```diff
@@ -5055,7 +5058,20 @@ def eval(self, expr, inplace=None, **kwargs):
         return self.map_partitions(M.eval, expr, meta=meta, inplace=inplace, **kwargs)
 
     @derived_from(pd.DataFrame)
-    def dropna(self, how="any", subset=None, thresh=None):
+    def dropna(self, how=pd_no_default, subset=None, thresh=pd_no_default):
+
+        if (
```
Comment (ncclementi) [inline dask/dataframe/core.py:5063]: This is causing a problem in the dropna test when we do `assert_eq(ddf.dropna(), df.dropna())` I think it's because it's trying to use pandas defaults. 

I'm not quite sure how to make things work on the tests to be consistent.

File: dask/dataframe/core.py [hunk 6]
```diff
@@ -5055,7 +5058,20 @@ def eval(self, expr, inplace=None, **kwargs):
...
+            not PANDAS_GT_150
+            and (how is not pd_no_default)
+            and (thresh is not pd_no_default)
+        ):
+            raise ValueError(
```
Comment (ncclementi) [inline dask/dataframe/core.py:5068]: Pandas error is a TypeError, but when we catch the errors in here https://github.com/dask/dask/blob/f7fd6c480478bdee5c7d6b8736ba69d5a2e918e0/dask/dataframe/utils.py#L215-L216 we use a ValueError so I'm using the same to be consistent.

File: dask/dataframe/core.py (hunk 3 above)
Comment (jrbourbeau) [reply dask/dataframe/core.py]: Yeah, I'd also drop the `not PANDAS_GT_150` part and just always raise that error appropriately

File: dask/dataframe/core.py (hunk 6 above)
Comment (ian-r-rose) [reply dask/dataframe/core.py:5062]: Interesting, I wonder why they raise a `TypeError`, `ValueError` seems more semantically correct to me as well.

I *might* prefer to be more consistent with pandas, so that user code could theoretically catch the same exception whether using dask or pandas, but I am not sure it's a huge deal either way

**Comment by ncclementi** [comment]
Thank you @jrbourbeau for that push

This failure seems unrelated, https://github.com/dask/dask/runs/7752422857?check_suite_focus=true
and the remaining tests on the upstream CI are the known failures.

File: dask/dataframe/core.py (hunk 6 above)
Comment (jrbourbeau) [reply dask/dataframe/core.py:5062]: Yeah, I had the same thought and preference as you. I'll go ahead and make this a `TypeError`. Like you said, probably not a big deal either way, but could lead to unexpected behavior if coming from `pandas`

File: dask/dataframe/core.py [hunk 7]
```diff
@@ -11,6 +11,7 @@
 
 import numpy as np
 import pandas as pd
+from pandas.api.extensions import no_default as pd_no_default
```
Comment (jrbourbeau) [inline dask/dataframe/core.py:14]: As is the changes here will pass the `no_default` from `pandas` down to the `.dropna(...)` method applied to each DataFrame partition. For Dask DataFrames that are backed by `pandas` DataFrames this should work totally fine, but my guess is this will break for `cudf`. cc @galipremsagar @quasiben for thoughts. 

xref https://github.com/dask/dask/issues/8645

File: dask/dataframe/tests/test_dataframe.py [hunk 8]
```diff
@@ -762,21 +762,17 @@ def test_dropna():
     )
 
     # threshold
-    assert_eq(df.dropna(thresh=None), df.loc[[20, 40]])
```
Comment (jrbourbeau) [inline dask/dataframe/tests/test_dataframe.py:765]: I removed a few asserts which were testing `pandas` but not `dask` (note these are using `df` instead of `ddf`). Not sure why these were added originally, but seem out of scope for our test suite.

**Review by jrbourbeau — COMMENTED** [review]
Alright, I think this is ready for another (possibly final) review pass. I opted to only forward only one of `how` or `thresh` in the underlying `map_partitions` call if one of them is specified (we raise an error early if both are specified). This will allow us to be more agnostic to the `pandas` version being used and also not break when operating on `cudf` DataFrames (I think). 

We should still push the conversation forward on https://github.com/dask/dask/issues/8645, but I don't think a conclusion there should block this PR.

File: dask/dataframe/core.py [hunk 9]
```diff
@@ -5055,10 +5055,23 @@ def eval(self, expr, inplace=None, **kwargs):
...
+        # These keywords are incompatible with each other.
+        # Don't allow them both to be set.
+        if how is not no_default and thresh is not no_default:
+            raise TypeError(
+                "You cannot set both the how and thresh arguments at the same time."
+            )
+
+        # Only specify `how` or `thresh` keyword if specified by the user
+        # so we utilize other `dropna` keyword defaults appropriately
+        kwargs = {"subset": subset}
```
Comment (ncclementi) [inline dask/dataframe/core.py:5068]: 😎

**Comment by ncclementi** [comment]
This looks good to me and tests related to this PR are happy.

**Comment by jrbourbeau** [comment]
Thanks @ncclementi @ian-r-rose
